ee/tabby-db/src/email_service_credential.rs

```diff
@@ -0,0 +1,89 @@
+use anyhow::Result;
+use rusqlite::OptionalExtension;
+
+use crate::DbConn;
+
+const EMAIL_CREDENTIAL_ROW_ID: i32 = 1;
+
+#[derive(Debug, PartialEq)]
+pub struct EmailServiceCredentialDAO {
+    pub smtp_username: String,
+    pub smtp_password: String,
+    pub smtp_server: String,
+}
+
+impl EmailServiceCredentialDAO {
+    fn new(smtp_username: String, smtp_password: String, smtp_server: String) -> Self {
+        Self {
+            smtp_username,
+            smtp_password,
+            smtp_server,
+        }
+    }
+}
+
+impl DbConn {
+    pub async fn read_email_service_credential(&self) -> Result<Option<EmailServiceCredentialDAO>> {
+        let res = self
+            .conn
+            .call(|c| {
+                Ok(c.query_row(
+                    "SELECT smtp_username, smtp_password, smtp_server FROM email_service_credential WHERE id=?",
+                    [EMAIL_CREDENTIAL_ROW_ID],
+                    |row| Ok(EmailServiceCredentialDAO::new(row.get(0)?, row.get(1)?, row.get(2)?)),
+                )
+                .optional())
+            })
+            .await?;
+        Ok(res?)
+    }
+
+    pub async fn update_email_service_credential(
+        &self,
+        creds: EmailServiceCredentialDAO,
+    ) -> Result<()> {
+        Ok(self
+            .conn
+            .call(move |c| {
+                c.execute("DELETE FROM email_service_credential", ())?;
```

Comment:
instead of using delete + insert, use upsert (insert ... on conflict set ...)

example: https://github.com/TabbyML/tabby/pull/1177/files#diff-44848e85056f66d2e1a18a09ddd88e861b9ee73f8509c77aacd1e9978c10299cL41

Reply:
this is not resolved

Reply:
Was it the where clause? Must have forgotten to put it back in after rewriting the query, I added it now.